embassy-nrf/src/time_driver.rs

```diff
@@ -83,194 +79,407 @@ impl AlarmState {
-        // safety: we're allowed to assume the AlarmState is created by us, and
-        // we never create one that's out of bounds.
-        unsafe { self.alarms.borrow(cs).get_unchecked(alarm.id() as usize) }
+    #[interrupt]
+    fn RTC1() {
+        DRIVER.on_interrupt()
     }
 
-    fn trigger_alarm(&self, n: usize, cs: CriticalSection) {
-        let r = rtc();
-        r.intenclr.write(|w| unsafe { w.bits(compare_n(n)) });
+    pub(crate) fn init(irq_prio: crate::interrupt::Priority) {
+        DRIVER.init(irq_prio)
+    }
+}
 
-        let alarm = &self.alarms.borrow(cs)[n];
-        alarm.timestamp.set(u64::MAX);
+#[cfg(feature = "time-driver-timer1")]
+pub mod driver_impl {
+    use crate::timer::Frequency;
 
-        // Call after clearing alarm, so the callback can set another alarm.
+    use super::*;
 
-        // safety:
-        // - we can ignore the possiblity of `f` being unset (null) because of the safety contract of `allocate_alarm`.
-        // - other than that we only store valid function pointers into alarm.callback
-        let f: fn(*mut ()) = unsafe { mem::transmute(alarm.callback.get()) };
-        f(alarm.ctx.get());
+    fn timer() -> &'static pac::timer0::RegisterBlock {
+        unsafe { &*pac::TIMER1::ptr() }
     }
-}
 
-impl Driver for RtcDriver {
-    fn now(&self) -> u64 {
-        // `period` MUST be read before `counter`, see comment at the top for details.
-        let period = self.period.load(Ordering::Relaxed);
-        compiler_fence(Ordering::Acquire);
-        let counter = rtc().counter.read().bits();
-        calc_now(period, counter)
+    /// Provides a Real Time Counter using the HFCLK. Uses
+    /// the TIMER peripheral in counter mode.
+    struct TimerDriver {
+        /// Number of 2^23 periods elapsed since boot.
+        period: AtomicU32,
+        alarm_count: AtomicU8,
+        /// Timestamp at which to fire alarm. u64::MAX if no alarm is scheduled.
+        alarms: Mutex<[AlarmState; ALARM_COUNT]>,
     }
 
-    unsafe fn allocate_alarm(&self) -> Option<AlarmHandle> {
-        let id = self
-            .alarm_count
-            .fetch_update(Ordering::AcqRel, Ordering::Acquire, |x| {
-                if x < ALARM_COUNT as u8 {
-                    Some(x + 1)
-                } else {
-                    None
+    const ALARM_STATE_NEW: AlarmState = AlarmState::new();
+    embassy::time_driver_impl!(static DRIVER: TimerDriver = TimerDriver {
+        period: AtomicU32::new(0),
+        alarm_count: AtomicU8::new(0),
+        alarms: Mutex::const_new(CriticalSectionRawMutex::new(), [ALARM_STATE_NEW; ALARM_COUNT]),
+    });
+
+    impl TimerDriver {
+        fn init(&'static self, irq_prio: crate::interrupt::Priority) {
+            let r = timer();
+            r.mode.write(|w| w.mode().timer());
+            r.prescaler
+                .write(|w| unsafe { w.prescaler().bits(Frequency::F62500Hz as u8) });
```

Comment:
Why this particular prescaler? I thought you wanted this to have higher precision. Higher tick rate means more preicision. With 1Mhz rate you can measure down to 1us of time. With 16Mhz you can measure down to 62.5ns.